spring-cloud-gateway-mvc/src/main/java/org/springframework/cloud/gateway/mvc/ProxyExchange.java

```diff
@@ -403,30 +404,27 @@ private void proxy() {
 
 	private void appendXForwarded(URI uri) {
 		// Append the legacy headers if they were already added upstream
-		String host = headers.getFirst("x-forwarded-host");
-		if (host == null) {
-			return;
+		appendXForwardedHeader("x-forwarded-host", () -> uri.getHost());
+		appendXForwardedHeader("x-forwarded-proto", () -> uri.getScheme());
+	}
+	
+	private void appendXForwardedHeader(String headerName, Supplier<String> uriValue) {
+		String existingValue = headers.getFirst(headerName);
+		if (existingValue != null) {
+			StringBuilder builder = new StringBuilder(existingValue);
+			builder.append(",").append(uriValue.get());
+			headers.set(headerName, builder.toString());
 		}
-		host = host + "," + uri.getHost();
-		headers.set("x-forwarded-host", host);
-		String proto = headers.getFirst("x-forwarded-proto");
-		if (proto == null) {
-			return;
-		}
-		proto = proto + "," + uri.getScheme();
-		headers.set("x-forwarded-proto", proto);
 	}
 
 	private void appendForwarded(URI uri) {
-		String forwarded = headers.getFirst("forwarded");
-		if (forwarded != null) {
-			forwarded = forwarded + ",";
-		}
-		else {
-			forwarded = "";
+		String forwarded = headers.getFirst("forwarded");		
+		StringBuilder builder = new StringBuilder();
+		if (StringUtils.hasText(forwarded)) {
```

Comment:
I would maintain the `null` check so that the same headers are added for the proxied request